multiqc/modules/mirtrace/mirtrace.py

```diff
@@ -0,0 +1,381 @@
+
+        # Write parsed data to a file
+        self.write_data_file(self.summary_data, 'multiqc_mirtrace_summary')
+        self.write_data_file(self.length_data, 'multiqc_mirtrace_length')
+        self.write_data_file(self.contamination_data, 'multiqc_mirtrace_contamination')
+        self.write_data_file(self.complexity_data, 'multiqc_mirtrace_complexity')
+
+        # Report sections
+        self.add_section (
+            name = 'QC Plot',
+            anchor = 'mirtrace_qc',
+            plot = self.mirtrace_qc_plot()
+        )
+
+        self.add_section (
+            name = 'Read Length Distribution',
+            anchor = 'mirtrace_length',
+            plot = self.mirtrace_length_plot()
+        )
+
+        self.add_section (
+            name = 'RNA Categories',
+            anchor = 'mirtrace_rna_categories',
+            plot = self.mirtrace_rna_categories()
+        )
+
+        self.add_section (
+            name = 'Contamination Check',
+            anchor = 'mirtrace_contamination_check',
+            plot = self.mirtrace_contamination_check()
+        )
+
+        self.add_section (
+            name = 'miRNA Complexity',
+            anchor = 'mirtrace_complexity',
+            plot = self.mirtrace_complexity_plot()
+        )
+
+    # Parse a miRTrace results.json file
+    def parse_summary(self, f):
+
+        try:
+            cdict = json.loads(f['f'])
+        except ValueError as e:
+            raise e
+
+        if 'results' in cdict.keys():
+            for record in cdict['results']:
+                s_name = self.clean_s_name(record['verbosename'], f['root'])
+                parsed_data = {}
+                parsed_data['filename'] = record['filename']
+                parsed_data['reads_total'] = record['stats']['allSeqsCount']
+                parsed_data['adapter_removed_length_ok'] = record['stats']['statsQC'][4]
+                parsed_data['adapter_not_detected'] = record['stats']['statsQC'][3]
+                parsed_data['length_shorter_than_18'] = record['stats']['statsQC'][2]
+                parsed_data['low_complexity'] = record['stats']['statsQC'][1]
+                parsed_data['low_phred'] = record['stats']['statsQC'][0]
+                parsed_data['reads_mirna'] = record['stats']['statsRNAType'][0]
+                parsed_data['reads_rrna'] = record['stats']['statsRNAType'][1]
+                parsed_data['reads_trna'] = record['stats']['statsRNAType'][2]
+                parsed_data['reads_artifact'] = record['stats']['statsRNAType'][3]
+                parsed_data['reads_unknown'] = record['stats']['statsRNAType'][4]
+                if s_name in self.summary_data:
+                    s_name = s_name + '_' + record['filename']
+                self.add_data_source(f, s_name)
+                self.summary_data[s_name] = parsed_data
+
+
+    # Parse a miRTrace mirtrace-stats-length.tsv file
+    def parse_length(self, f):
+        try:
+            with open(f['fn'], 'rb') as file:
```

Comment:
If you use `self.find_log_files('mirtrace/length', filehandles=True):` then MultiQC will give you already opened file handles. Then you can simplify this code a little by removing the `with` statement:

```python
try:
    reader = csv.reader(f['f'], delimiter="\t")